# Hacker News - posts analysis

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

The data set consists of aproximately 20,000 rows. Below are descriptions of the columns:

-id: The unique identifier from Hacker News for the post
-title: The title of the post
-url: The URL that the posts links to, if it the post has a URL
-num_points: The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
-num_comments: The number of comments that were made on the post
-author: The username of the person who submitted the post
-created_at: The date and time at which the post was submitted

There are two different types of titles that begin with either Ask HN or Shown HN. Users submit Ask HN posts to ask the Hacker News community a specific question. Users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

In this project I will try to compare two types of posts to determine the following:

-Do Ask HN or Show HN receive more comments on average?

-Do posts created at a certain time receive more comments on average?

In [9]:
from csv import reader

opened_file = open('hacker_news.csv')

read_file = reader(opened_file)

hn = list(read_file)

hn_header = hn[0]

hn_data = hn[1:]

In [10]:
print(hn[0])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [13]:
print(hn_data[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


### Filter out posts.

In [17]:
ask_posts = []
show_posts = []
other_posts = []


for row in hn_data:
    title = row[1]
    
    if title.lower().startswith('ask hn'):
        ask_posts.append(row)
    elif title.lower().startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


In [18]:
print(ask_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


In [19]:
print(show_posts[:5])

[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']]


### Trying to determine which posts receive more comments on average.

In [20]:
total_ask_comments = 0

for post in ask_posts:
    num_comments = int(post[4])
    total_ask_comments += num_comments
    
avg_ask_comments = total_ask_comments/len(ask_posts)

print(avg_ask_comments)

14.038417431192661


In [21]:
total_show_comments =0

for post in show_posts:
    num_comments = int(post[4])
    total_show_comments += num_comments
    
avg_show_comments = total_show_comments/len(show_posts)
print(avg_show_comments)

10.31669535283993


In average ask posts receive more comments approximately 14 on one comment. Whereas show posts receive approx 10. Due to ask posts receive moe comments I will focus on this posts to do my analysis. 

### Calculating amount of ask posts and comments by hour created.

In [22]:
import datetime as dt

In [31]:
result_list = []

for row in ask_posts:
    result_list.append([row[6],int(row[4])])

counts_by_hour = {} # number of ask posts created during each hour of the day
comments_by_hour = {} # corresponding number of comments ask posts created at each hour received
date_format = "%m/%d/%Y %H:%M"

for item in result_list:
    hour = item[0]
    comment = item[1]
    time = dt.datetime.strptime(hour,date_format).strftime('%H')
    
    if time not in counts_by_hour:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment
    else:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment
    

In [32]:
print(counts_by_hour)

{'09': 45, '23': 68, '07': 34, '15': 116, '08': 48, '10': 59, '03': 54, '16': 108, '13': 85, '19': 110, '20': 80, '02': 58, '00': 55, '12': 73, '14': 107, '01': 60, '18': 109, '11': 58, '17': 100, '22': 71, '21': 109, '05': 46, '06': 44, '04': 47}


In [33]:
print(comments_by_hour)

{'09': 251, '23': 543, '07': 267, '15': 4477, '08': 492, '10': 793, '03': 421, '16': 1814, '13': 1253, '19': 1188, '20': 1722, '02': 1381, '00': 447, '12': 687, '14': 1416, '01': 683, '18': 1439, '11': 641, '17': 1146, '22': 479, '21': 1745, '05': 464, '06': 397, '04': 337}


###  Calculating average number of comments per post (for posts created during each hour of the day)

In [34]:
avg_by_hour = []

for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour]/counts_by_hour[hour]])

In [37]:
avg_by_hour

[['09', 5.5777777777777775],
 ['23', 7.985294117647059],
 ['07', 7.852941176470588],
 ['15', 38.5948275862069],
 ['08', 10.25],
 ['10', 13.440677966101696],
 ['03', 7.796296296296297],
 ['16', 16.796296296296298],
 ['13', 14.741176470588234],
 ['19', 10.8],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['00', 8.127272727272727],
 ['12', 9.41095890410959],
 ['14', 13.233644859813085],
 ['01', 11.383333333333333],
 ['18', 13.20183486238532],
 ['11', 11.051724137931034],
 ['17', 11.46],
 ['22', 6.746478873239437],
 ['21', 16.009174311926607],
 ['05', 10.08695652173913],
 ['06', 9.022727272727273],
 ['04', 7.170212765957447]]

In [38]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
    
print(swap_avg_by_hour)

[[5.5777777777777775, '09'], [7.985294117647059, '23'], [7.852941176470588, '07'], [38.5948275862069, '15'], [10.25, '08'], [13.440677966101696, '10'], [7.796296296296297, '03'], [16.796296296296298, '16'], [14.741176470588234, '13'], [10.8, '19'], [21.525, '20'], [23.810344827586206, '02'], [8.127272727272727, '00'], [9.41095890410959, '12'], [13.233644859813085, '14'], [11.383333333333333, '01'], [13.20183486238532, '18'], [11.051724137931034, '11'], [11.46, '17'], [6.746478873239437, '22'], [16.009174311926607, '21'], [10.08695652173913, '05'], [9.022727272727273, '06'], [7.170212765957447, '04']]


In [40]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

In [41]:
sorted_swap

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [43]:
print('Top 5 Hours for Ask Posts Comments')

for avg,hour in sorted_swap[:5]:
    print('{}: {:.2f} average comments per post'.format(dt.datetime.strptime(hour,'%H').strftime('%H:%M'),avg))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

## Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on this analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.